### ISSUE TITLE
Additional option for window/showTextDocument

### ISSUE THREAD

**Issue by tgodzik** [microsoft/language-server-protocol#1140]
We already have a very similar command that we send to the client to display a particular document:

https://github.com/scalameta/metals/blob/9e2a418b210518567fe51483167658466c9e256c/metals/src/main/scala/scala/meta/internal/metals/ClientCommands.scala#L108

we also added an additional boolean option `otherWindow` which we use to specify that the document should not be opened in the focused currently window/tab.

I wonder if it would make sense to add that property to `window/showTextDocument` 

This is especially useful for us when navigating over a stacktrace, which makes it possible to quickly jump around different classes without the need to go back to the stacktrace.

**Comment by rchl** [comment]
`otherWindow` sounds too specific to me. Why? because editors can provide means to open something in a way that doesn't affect the current view and it doesn't need to be a separate window specifically.

For example, in Sublime Text, you can trigger the opening of a separate view in a side-by-side manner. I suppose that in VScode you could also do something similar or use a panel.

So I think the naming should be more generic. Not sure what exactly. `keepCurrent`? I guess the current is always "kept` so maybe not ideal? Is the idea that the current view stays in focus or not necessarily?

**Comment by tgodzik** [comment]
@rchl you're right,  `otherWindow` might be too specific. It's just what we come up with and naming is hard :sweat_smile: 

> Is the idea that the current view stays in focus or not necessarily?

Yes, the idea would be to keep the current view in focus and show the new document outside it if possible. If an editor does not support side by side presentation they should just ignore it I guess.

So maybe something like: `keepFocus` or `showOnSide` ?

**Comment by rchl** [comment]
I kinda also see problems with implementing the "keep focus" part of that as I think it wouldn't be possible to open a new file without changing focus in some editors.

**Comment by tgodzik** [comment]
> I kinda also see problems with implementing the "keep focus" part of that as I think it wouldn't be possible to open a new file without changing focus in some editors.

Focus might not be that necessary, the important thing would be to keep the current document visible. I understand if that might be too niche to include in the protocol.

**Comment by rchl** [comment]
OK, but I don't see how you can guarantee that while opening a new window either. It will likely cover the current window.

**Comment by tgodzik** [comment]
> OK, but I don't see how you can guarantee that while opening a new window either. It will likely cover the current window.

Not sure I understand. If the server did send the additional option `keepCurrentDocumentVisible` then it would be client's responsibility to guarantee that if it's possible for that client.

**Comment by rchl** [comment]
Just wondering how was your original solution that uses `otherWindow` been able to achieve the goal of keeping the current view in focus and in view if it's opening a new window.

**Comment by tgodzik** [comment]
For example for Visual studio Code extension we did it like this:

https://github.com/scalameta/metals-vscode/blob/master/src/extension.ts#L828

**Comment by dbaeumer** [comment]
I am not a fan of such a feature since it depends on the capabilities of the client (e.g. multiple editor columns, ...).

**Comment by tgodzik** [comment]
> I am not a fan of such a feature since it depends on the capabilities of the client (e.g. multiple editor columns, ...).

One could argue that most features really depend on capabilities of the client, but I do understand what you mean here. We can always go around it with custom commands.

**Comment by dbaeumer** [comment]
Yes, but these capabilities are usually not UI capabilities

**Comment by tgodzik** [comment]
> Yes, but these capabilities are usually not UI capabilities

I wonder if it would make sense to have some officially supported extensions to LSP that are not a part of base spec, which would differ in the UI capabilities. This could make supporting multiple clients easier, but just throwing it out there.

**Comment by dbaeumer** [comment]
I think that would be a better direction than having it in the base spec.
